cmd/srcd/cmd/sql.go

```diff
@@ -43,98 +46,64 @@ var sqlCmd = &cobra.Command{
 
 		var query string
 		if len(args) == 1 {
-			query = args[0]
+			query = strings.TrimSpace(args[0])
 		}
 
 		client, err := daemon.Client()
 		if err != nil {
 			fatal(err, "could not get daemon client")
 		}
 
-		timeout := 3 * time.Second
-		started := logAfterTimeoutWithServerLogs("this is taking a while, "+
-			"if this is the first time you launch sql client, "+
-			"it might take a few more minutes while we install all the required images",
-			timeout)
-
-		// Might have to pull some images
-		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
-		_, err = client.StartComponent(ctx, &api.StartComponentRequest{
-			Name: components.Gitbase.Name,
-		})
-		started()
-		cancel()
-
-		if err != nil {
-			fatal(err, "could not start gitbase")
-		}
-
-		connReady := logAfterTimeoutWithSpinner("waiting for gitbase to be ready", timeout, 0)
+		startGitbaseWithClient(client)
+		connReady := logAfterTimeoutWithSpinner("waiting for gitbase to be ready", 5*time.Second, 0)
 		err = ensureConnReady(client)
 		connReady()
 		if err != nil {
 			fatal(err, "could not connect to gitbase")
 		}
 
-		if strings.TrimSpace(query) == "" {
-			if err := repl(client); err != nil {
-				log.Fatal(err)
+		// Support piping
+		// TODO(@smacker): not the most optimal solution
+		// it would read all input into memory first and only then send to gitbase
+		// it must be possible to pipe and running mysql-cli with -B flag
+		// but it would change current client behaviour
+		fi, _ := os.Stdin.Stat()
+		if (fi.Mode() & os.ModeCharDevice) == 0 {
+			b, err := ioutil.ReadAll(os.Stdin)
+			if err != nil {
+				fatal(err, "could not read input")
 			}
-			return nil
-		}
 
-		if err := runQuery(client, query); err != nil {
-			log.Fatal(err)
+			query = string(b)
 		}
-		return nil
-	},
-}
-
-func repl(client api.EngineClient) error {
-	rl, err := readline.NewEx(&readline.Config{
-		Prompt: "gitbase> ",
-		Stdin:  os.Stdin,
-		Stderr: os.Stderr,
-		Stdout: os.Stdout,
-	})
-	if err != nil {
-		return err
-	}
-	defer rl.Close()
 
-	for {
-		// read until you get a trailing ';'.
-		var lines []string
-		for {
-			line, err := rl.Readline()
-			if err != nil {
-				if err != io.EOF {
-					log.Fatalf("could not read line: %v", err)
-				}
-				return nil
-			}
-			line = strings.TrimSpace(line)
-			lines = append(lines, line)
-			if strings.HasSuffix(line, ";") {
-				rl.SetPrompt("gitbase> ")
-				break
-			}
-			rl.SetPrompt("      -> ")
+		resp, exit, err := runMysqlCli(context.Background(), query)
+		if err != nil {
+			fatal(err, "could not run mysql client")
 		}
+		defer resp.Close()
+		defer stopMysqlClient()
+
+		// in case of Ctrl-C or kill defer wouldn't work
+		ch := make(chan os.Signal, 1)
```

Comment:
Maybe we could unify this with the logic in `attachStdio`? I was thinking about sending `nil` (or an error identifying the signal) to `ouputDone` or `inputDone` when one of the two signals arrive.

Reply:
we can't put this logic into `attachStdio` because we need to handle signal correctly for the case when we don't attach stdin too. Example: you run `select * from files` which takes a lot of time, Ctrl-C should stop the query and remove container correctly. Instead of writing additional code for this case we can handle it using this function in both cases with or without stdin.

Reply:
ok! gonna approved then! thank!